content/en/blog/_posts/2023-04-11-kmsv2-beta.md

```diff
@@ -0,0 +1,188 @@
+    participant external_kms
+    alt Generate DEK at startup
+        Note over kube_api_server,external_kms: Refer to Generate Data Encryption Key (DEK) diagram for details
+    end
+    user->>kube_api_server: create/update resource that's to be encrypted
+    kube_api_server->>kube_api_server: encrypt resource with DEK
+    kube_api_server->>etcd: store encrypted object
+```
+-->
+
+{{< figure src="/images/blog/2023-04-11-kubernetes-1.27-kmsv2-beta/kubernetes-1.27-encryption.svg" alt="Sequence diagram for KMSv2 beta Encrypt" class="diagram-large" >}}
+
+#### Decrypt Request
+
+<!-- source
+```mermaid
+%%{init:{"theme":"neutral", "sequence": {"mirrorActors":true},
+    "themeVariables": {
+        "actorBkg":"royalblue",
+        "actorTextColor":"white"
+}}}%%
+
+sequenceDiagram
+    participant user
+    participant kube_api_server
+    participant kms_plugin
+    participant external_kms
+    participant etcd
+    user->>kube_api_server: get/list resource that's encrypted
+    kube_api_server->>etcd: get encrypted resource
+    etcd->>kube_api_server: encrypted resource
+    alt Encrypted DEK not in cache
+        kube_api_server->>kms_plugin: decrypt request
+        kms_plugin->>external_kms: decrypt DEK with remote KEK
+        external_kms->>kms_plugin: decrypted DEK
+        kms_plugin->>kube_api_server: return decrypted DEK
+        kube_api_server->>kube_api_server: cache decrypted DEK
+    end
+    kube_api_server->>kube_api_server: decrypt resource with DEK
+    kube_api_server->>user: return decrypted resource
+```
+-->
+
+{{< figure src="/images/blog/2023-04-11-kubernetes-1.27-kmsv2-beta/kubernetes-1.27-decryption.svg" alt="Sequence diagram for KMSv2 beta Decrypt" class="diagram-large" >}}
+
+#### Status Request
+
+<!-- source
+```mermaid
+%%{init:{"theme":"neutral", "sequence": {"mirrorActors":true},
+    "themeVariables": {
+        "actorBkg":"royalblue",
+        "actorTextColor":"white"
+}}}%%
+
+sequenceDiagram
+    participant kube_api_server
+    participant kms_plugin
+    participant external_kms
+    alt Generate DEK at startup
+        Note over kube_api_server,external_kms: Refer to Generate Data Encryption Key (DEK) diagram for details
+    end
+    loop every minute (or every 10s if error or unhealthy)
+        kube_api_server->>kms_plugin: status request
+        kms_plugin->>external_kms: validate remote KEK
+        external_kms->>kms_plugin: KEK status
+        kms_plugin->>kube_api_server: return status response <br/> {"healthz": "ok", key_id: "<remote KEK ID>", "version": "v2beta1"}
+        alt KEK rotation detected (key_id changed), rotate DEK
+            Note over kube_api_server,external_kms: Refer to Generate Data Encryption Key (DEK) diagram for details
+        end
+    end
+```
+-->
+
+{{< figure src="/images/blog/2023-04-11-kubernetes-1.27-kmsv2-beta/kubernetes-1.27-status.svg" alt="Sequence diagram for KMSv2 beta Status" class="diagram-large" >}}
+
+#### Generate Data Encryption Key (DEK)
```

Comment:
There's nothing in this diagram that shows the reuse of DEK and a new DEK is only generated on server startup and a KEK rotation. I think this is the big value add for v2beta1.

Reply:
`Generate Data Encryption Key (DEK)` is to only show what the flow of events is when a DEK is generated (during startup or rotation of KEK).  This was used in multiple places, so moved it a single diagram and this is referenced in the ones above. 

The [sequence diagram for Encrypt](https://deploy-preview-40194--kubernetes-io-main-staging.netlify.app/blog/2023/04/11/kms-v2-moves-to-beta/#encrypt-request) shows the DEK is only generated at startup, the [diagram for status](https://deploy-preview-40194--kubernetes-io-main-staging.netlify.app/blog/2023/04/11/kms-v2-moves-to-beta/#status-request) shows in addition to startup, the DEK being generated when the KEK is rotated.

Reply:
Ah ok. my eyes went straight to the gray box and didnt see the `[]` text above it.